fem/bilininteg_gradient.cpp

```diff
@@ -0,0 +1,800 @@
+         {
+            grad[qy][qx][0] = 0.0;
+            grad[qy][qx][1] = 0.0;
+         }
+      }
+      for (int dy = 0; dy < TR_D1D; ++dy)
+      {
+         double gradX[max_Q1D][VDIM];
+         for (int qx = 0; qx < Q1D; ++qx)
+         {
+            gradX[qx][0] = 0.0;
+            gradX[qx][1] = 0.0;
+         }
+         for (int dx = 0; dx < TR_D1D; ++dx)
+         {
+            const double s = x(dx,dy,e);
+            for (int qx = 0; qx < Q1D; ++qx)
+            {
+               gradX[qx][0] += s * G(qx,dx);
+               gradX[qx][1] += s * B(qx,dx);
+            }
+         }
+         for (int qy = 0; qy < Q1D; ++qy)
+         {
+            const double wy  = B(qy,dy);
+            const double wDy = G(qy,dy);
+            for (int qx = 0; qx < Q1D; ++qx)
+            {
+               grad[qy][qx][0] += gradX[qx][0] * wy;
+               grad[qy][qx][1] += gradX[qx][1] * wDy;
+            }
+         }
+      }
+      // We've now calculated grad(p) = [Dxy, xDy] in plane
+      for (int qy = 0; qy < Q1D; ++qy)
+      {
+         for (int qx = 0; qx < Q1D; ++qx)
+         {
+            const int q = qx + qy * Q1D;
+            const double gradX = grad[qy][qx][0];
+            const double gradY = grad[qy][qx][1];
+
+            grad[qy][qx][0] = gradX*op(q,0,0,e) + gradY*op(q,1,0,e);
+            grad[qy][qx][1] = gradX*op(q,0,1,e) + gradY*op(q,1,1,e);
+         }
+      }
+      // We've now calculated grad = grad p * op
+      for (int qy = 0; qy < Q1D; ++qy)
+      {
+         double opX[max_TE_D1D][VDIM];
+         for (int dx = 0; dx < TE_D1D; ++dx)
+         {
+            opX[dx][0] = 0.0;
+            opX[dx][1] = 0.0;
+            for (int qx = 0; qx < Q1D; ++qx)
+            {
+               opX[dx][0] += Bt(dx,qx)*grad[qy][qx][0];
+               opX[dx][1] += Bt(dx,qx)*grad[qy][qx][1];
+            }
+         }
+         for (int dy = 0; dy < TE_D1D; ++dy)
+         {
+            for (int dx = 0; dx < TE_D1D; ++dx)
+            {
+               y(dx,dy,0,e) += Bt(dy,qy)*opX[dx][0];
+               y(dx,dy,1,e) += Bt(dy,qy)*opX[dx][1];
+            }
+         }
+      }
+      // We've now calculated y = u * grad
+   });
+
+}
+
+// Shared memory PA Gradient Apply 2D kernel
+template<const int T_TR_D1D = 0, const int T_TE_D1D = 0, const int T_Q1D = 0,
+         const int T_NBZ = 0>
+static void SmemPAGradientApply2D(const int NE,
+                                  const Array<double> &_b,
+                                  const Array<double> &_g,
+                                  const Array<double> &_bt,
+                                  const Vector &_op,
+                                  const Vector &_x,
+                                  Vector &_y,
+                                  const int tr_d1d = 0,
+                                  const int te_d1d = 0,
+                                  const int q1d = 0)
+{
+   // TODO
+   MFEM_ASSERT(false, "SHARED MEM NOT PROGRAMMED YET");
+}
```

Comment:
Better to remove the placeholder for the shared memory version until it's implemented